feat: poll for new alerts every 5 seconds

PureTryOut opened the pull request:
This removes the need for the user to manually refresh the list as it will now happen automatically instead.

nucleus-ffm reviewed (commented):
I noticed that these changes break some features: 

- No error message will be shown if the server request fails. It just looks like there are no alerts
- It looks like it does not pull every 5s but every 10s. 
- The read status of alerts is not stored. it is always shown as unread
- The Update thread of alerts is broken. All alerts are shown even if there is an update for them. But this may have broken already before this PR

Currently, we do not use any caching or etags. So if there is an alert with 1MB size and we fetch that every 5s, this results in 60MB traffic in 5 minutes. So we definitely need some caching and etag mechanism to avoid having too much load on the server and too much traffic on the client. 

Also, it is not visible whether the update loop is working or not. If this should break for some reason, it is not possible to notice that. 

Is there a way to turn off the polling in the dev environment? In some situations, you may not want to poll all the time but instead in a more controlled manner.

PureTryOut commented:
> Currently, we do not use any caching or etags. So if there is an alert with 1MB size and we fetch that every 5s, this results in 60MB traffic in 5 minutes. So we definitely need some caching and etag mechanism to avoid having too much load on the server and too much traffic on the client.

True. However the problem imo is not necessarily the lack of caching but more so that we poll for the entirety of the alert detail rather than just what alerts have changed. Right now `alertsForPlaceFutureProvider` gets both what alerts are there and then what kind of detail they have. That should probably be changed to just get the list of relevant alerts, compare that with a local list of alerts we already have and then only fetch detail for the new ones. That saves a bunch of data because suddenly we're only getting a few strings per poll rather than entire JSON structures.

> Is there a way to turn off the polling in the dev environment? In some situations, you may not want to poll all the time but instead in a more controlled manner.

Not at the moment, but I also don't see why this would be good to have? I don't want my dev environment to be _too_ different from "production" because it'll not be representative of the actual thing and I'll miss potential bugs and solutions.

> Also, it is not visible whether the update loop is working or not. If this should break for some reason, it is not possible to notice that.

I suppose so. I guess I could keep track of an error state somewhere.

PureTryOut commented on lib/services/warnings.dart:
```diff
@@ -5,20 +5,70 @@ import 'package:foss_warn/class/class_warn_message.dart';
+/// It polls for new alerts and merges the result with any locally processed/modified alerts.
+/// Any processing on alerts has to be done through [processedAlertsProvider].
+final alertsProvider = Provider<List<WarnMessage>>((ref) {
+  ref.listen(tickingChangeProvider(100), (_, event) {
```
I changed it back to 5 seconds btw.

PureTryOut commented:
We now only fetch the alert ID's every 5 seconds. We only fetch details if we detect any alert we didn't know about yet before.
Any errors occurring while fetching is also shown in the UI.

nucleus-ffm commented on lib/views/add_my_place_with_map_view.dart:
```diff
@@ -565,12 +563,10 @@ class _AddMyPlaceWithMapViewState extends ConsumerState<AddMyPlaceWithMapView> {
                                 );
 
                                 setState(() {
```
Why do we still need a setState here? Isn't the rebuilding completely handled by the provider?

PureTryOut replied:
Leftover 🤷 I've removed it.

nucleus-ffm commented on lib/views/map_view.dart:
```diff
@@ -83,7 +83,7 @@ class _MapViewState extends ConsumerState<MapView> {
   @override
   Widget build(BuildContext context) {
-    var warnings = ref.watch(warningsProvider);
+    var alerts = ref.watch(processedAlertsProvider);
```
Does processedAlertsProvider only contain alerts that are already read? The Map is always empty and only displays alerts after they are marked as read.

nucleus-ffm commented on lib/views/my_places_view.dart:
```diff
@@ -42,121 +35,69 @@ class _MyPlacesState extends ConsumerState<MyPlacesView>
+                children: [
+                  if (appState.error ||
+                      userPreferences.areWarningsFromCache) ...[
                     const ConnectionError(),
```
Can we use the same logic here as in the warning view? Currently, if the fetch fails, we only display an error message in the warnings view. Additionally, it would actually be great if we could display a different dialog if we just don't have an internet connection right now. Displaying the error dialog may result in a wrong perception of the problem.  

```
if (alertsSnapshot.hasError) ...[
          const ConnectionError(),
        ],
```

nucleus-ffm reviewed (approved):
Ok map is working as expected again. But the read status is completely broken again. I cannot even mark one alert as read. But it would be fine for me if we fix the entire read-status thing in another PR.

PureTryOut commented:
Eh I added something last minute but that clearly made it worse. I reverted that for now so marking as read works again. Thanks for approving, we'll iterate on this further!
